quilt_lite/quilt_lite/graphql.py

```diff
@@ -0,0 +1,282 @@
+
+        if not isinstance(entry, quilt3.packages.PackageEntry):
+            return None
+        return PackageFileWrapper(path, entry)
+
+    def accessCounts(self, *_, **__):
+        return None
+
+class RevisionListWrapper:
+    def __init__(self, bucket: str, name: str):
+        self.bucket = bucket
+        self.name = name
+
+    @functools.cached_property
+    def _package_versions(self):
+        # XXX: for some reason it's sooooooo slow
+        return dict(quilt3.list_package_versions(self.name, f"s3://{self.bucket}"))
+
+    @functools.cached_property
+    def _revisions_by_hash(self):
+        by_hash = {}
+        for tag, hash in self._package_versions.items():
+            if hash in by_hash:
+                by_hash[hash].append(tag)
+            else:
+                by_hash[hash] = [tag]
+        return by_hash
+
+    @functools.cached_property
+    def _revision_wrappers(self):
+        #XXX: order?
+        wrappers = [RevisionWrapper(bucket=self.bucket, name=self.name, hash=hash, tags=tags) for hash, tags in self._revisions_by_hash.items()]
+        return wrappers
+
+    @functools.cached_property
+    def total(self):
+        print('versions', self._revisions_by_hash)
+        return len(self._revisions_by_hash)
+
+    def page(self, *_, number: int, perPage: int):
+        offset = (number - 1) * perPage
+        return self._revision_wrappers[offset:offset + perPage]
+
+
+POINTER_RE = re.compile("^1[0-9]{9}$")
+
+
+def is_pointer(hash_or_tag: str):
+    if hash_or_tag == "latest": return True
+    if POINTER_RE.match(hash_or_tag): return True
+    return False
+
+class PackageWrapper:
+    def __init__(self, bucket: str, name: str):
+        self.bucket = bucket
+        self.name = name
+
+    @functools.cached_property
+    def _browse(self):
+        return quilt3.Package.browse(self.name, f"s3://{self.bucket}")
+
+    @functools.cached_property
+    def modified(self):
+        # TODO: figure out where to find mtime (manifest timestamp maybe?)
+        return datetime.datetime.fromisoformat("2021-10-21")
+
+    @functools.cached_property
+    def revisions(self):
+        return RevisionListWrapper(self.bucket, self.name)
+
+    def revision(self, *_, hashOrTag: str):
+        registry = quilt3.backends.get_package_registry(f"s3://{self.bucket}")
+        hash = (
+            quilt3.data_transfer.get_bytes(registry.pointer_pk(self.name, hashOrTag)).decode()
+            if is_pointer(hashOrTag) else
+            registry.resolve_top_hash(self.name, hashOrTag)
+        )
+        return RevisionWrapper(bucket=self.bucket, name=self.name, hash=hash)
+
+    def accessCounts(self, *_, **__):
+        return None
+
+
+class PackageListWrapper:
+    def __init__(self, bucket: str, filter: T.Optional[str] = None):
+        self.bucket = bucket
+        self.filter = filter
+
+    @functools.cached_property
+    def _package_list(self):
+        return quilt3.list_packages(f"s3://{self.bucket}")
```

Comment:
It looks like a bad idea to cache iterator, since it's mutable and you can't get results from it twice.